opencti-platform/opencti-graphql/src/parser/csv-bundler.ts

```diff
@@ -71,3 +74,127 @@ export const bundleProcess = async (
   // Build and return the result
   return bundleBuilder.build();
 };
+
+/**
+ * Generate stix bundles. Try to put as much data as possible in the first bundle.
+ * Creates a new bundle when a data already exists with the same stixId but the content is different from previously
+ *  to allow upsert inside a same CSV file.
+ * @param context
+ * @param user
+ * @param lines
+ * @param mapper
+ * @param opts
+ */
+export const bundleAllowUpsertProcess = async (
+  context: AuthContext,
+  user: AuthUser,
+  lines: string[],
+  mapper: CsvMapperParsed,
+  opts: BundleProcessOpts = {}
+) => {
+  const { entity, maxRecordNumber } = opts;
+  const allBundles: BundleBuilder[] = [];
+  allBundles.push(new BundleBuilder());
+  const rawRecords = await parsingProcess(lines, mapper.separator, mapper.skipLineChar);
+  const records = maxRecordNumber ? rawRecords.slice(0, maxRecordNumber) : rawRecords;
+  const refEntities = await handleRefEntities(context, user, mapper);
+  if (records) {
+    for (let rec = 0; rec < records.length; rec += 1) {
+      const record = records[rec];
+      const isEmptyLine = record.length === 1 && isEmptyField(record[0]);
+      if (!isEmptyLine) {
+        try {
+          // Compute input by representation
+          const inputs = await mappingProcess(context, user, mapper, record, refEntities);
+          // Remove inline elements
+          const withoutInlineInputs = inputs.filter((input) => !inlineEntityTypes.includes(input.entity_type as string));
+          // Transform entity to stix
+          const stixObjects = withoutInlineInputs.map((input) => {
+            const stixObject = convertStoreToStix(input as unknown as StoreCommon);
+            stixObject.extensions[STIX_EXT_OCTI].converter_csv = record.join(mapper.separator);
+            return stixObject;
+          });
+
+          // Add to bundle
+          let added: boolean = false;
+          let bundleIndex = 0;
+          while (!added && bundleIndex < allBundles.length) {
+            if (allBundles[bundleIndex].canAddObjects(stixObjects)) {
+              allBundles[0].addObjects(stixObjects);
+              added = true;
+            }
+            bundleIndex += 1;
+          }
+
+          if (!added) {
+            const nextBuilder = new BundleBuilder();
+            nextBuilder.addObjects(stixObjects);
+            allBundles.push(nextBuilder);
+          }
+        } catch (e) {
+          logApp.error(e);
+        }
+      }
+    }
+  }
+  // Handle container
+  if (entity && isStixDomainObjectContainer(entity.entity_type)) {
+    for (let i = 0; i < allBundles.length; i += 1) {
+      const currentBundle = allBundles[i];
+      const refs = currentBundle.ids();
+      const stixEntity = { ...convertStoreToStix(entity), [objects.stixName]: refs };
+      currentBundle.addObject(stixEntity);
+    }
+  }
```

Comment:
We maybe need to check that the entity itself isn't already contained in the bundle also?